### ISSUE TITLE
SPEC - document the order of operations for updates

### ISSUE THREAD

**Issue by amorton** [stargate/jsonapi#132]
clarify...

- The order we process the operations in the update document. We will process them in document order for the update doc
e.g. 

```
{
  "$unset" : {...}, 
  "$set" : {...}, 
}
```

the $unset op runs before the $set 

- For the operations, we process the updates in string field order. This has implications for arrays with indexing. 

e.g.
`{"$set": {"foo" : 1, "bar" : 2}}` sets the bar field to 2 before setting foo to 1

**Comment by tatu-at-datastax** [comment]
Also worth noting: it is illegal to refer to same field from both `$set` and `$unset` -- this is being verified by implementation.

**Comment by vkarpov15** [comment]
@tatu-at-datastax good point. I'd say order doesn't matter very much, because MongoDB throws an error if you attempt to update the same field with 2 different update operators, or attempt to update a path with one update operator and a subpath with a different update operator. So you don't have to worry about what order to apply `{ $set: { foo: { bar: 'baz' } }, $unset: { foo: 1 } }`, IMO that should be an error.

**Comment by tatu-at-datastax** [comment]
@vkarpov15 Ok. Exact path checking between $set and $unset is implemented, but sub-path is quite a bit of more work to verify. Good to know Mongo prevents that as well, I'll have to to double-check. Thank you for pointing this out!
(technically one could just apply changes in order and accept changes because setting/unsetting of not-yet-existing paths are acceptable operations, as long as there's no attempts to change property of Array).

**Comment by amorton** [comment]
> For us I think document order, as explained, makes most sense: it is simple to understand and specify by users.
@tatu-at-datastax  the order of field updates inside on operation needs to be lexigraphical / string order not document order as in the example at the top.

**Comment by tatu-at-datastax** [comment]
Ok, I will create a new issue for this then.
